File: src/core/examples.ts
import { loadSpec } from "./spec";
import type { ExampleTranslation } from "./types";

export function listExamples(): ExampleTranslation[] {
  return loadSpec().examples.examples;
}

export function findExample(query: string): ExampleTranslation | undefined {
  const examples = listExamples();
  const number = Number(query);
  if (Number.isInteger(number)) {
    return examples.find((example) => example.number === number);
  }
  const lowered = query.toLowerCase();
  return examples.find((example) => example.english.toLowerCase().includes(lowered));
}

File: crates/ethra-language/src/examples.rs
use crate::spec::load_spec;
use crate::types::ExampleTranslation;
use anyhow::Result;

pub fn list_examples() -> Result<Vec<ExampleTranslation>> {
    Ok(load_spec()?.examples.examples.clone())
}

pub fn find_example(query: &str) -> Result<Option<ExampleTranslation>> {
    let examples = list_examples()?;
    if let Ok(number) = query.parse::<i64>() {
        return Ok(examples
            .into_iter()
            .find(|example| example.number == number));
    }
    let lowered = query.to_lowercase();
    Ok(examples
        .into_iter()
        .find(|example| example.english.to_lowercase().contains(&lowered)))
}
